pkg/kube/krt/dynamicjoin.go

```diff
@@ -0,0 +1,436 @@
+	eventType       indexEventType
+	collectionName  string
+	collectionValue internalCollection[T]
+}
+
+// nolint: unused // (not true)
+type dynamicJoinIndexer struct {
+	indexers map[string]kclient.RawIndexer
+	sync.RWMutex
+}
+
+// nolint: unused // (not true)
+func (j *dynamicJoinIndexer) Lookup(key string) []any {
+	var res []any
+	first := true
+	j.RLock()
+	defer j.RUnlock() // keithmattix: we're probably fine to defer as long as we don't have nested dynamic indexers
+	for _, i := range j.indexers {
+		l := i.Lookup(key)
+		if len(l) > 0 && first {
+			// Optimization: re-use the first returned slice
+			res = l
+			first = false
+		} else {
+			res = append(res, l...)
+		}
+	}
+	return res
+}
+
+// nolint: unused // (not true, its to implement an interface)
+func (j *dynamicjoin[T]) index(extract func(o T) []string) kclient.RawIndexer {
+	j.Lock() // Take a write lock since we're also updating the collection change handlers
+	defer j.Unlock()
+	ji := &dynamicJoinIndexer{indexers: make(map[string]kclient.RawIndexer, len(j.collectionsByKey))}
+	for _, c := range maps.SeqStable(j.collectionsByKey) {
+		ji.indexers[c.name()] = c.index(extract)
+	}
+	j.registerCollectionChangeHandlerLocked(func(e collectionChangeEvent[T]) {
+		ji.Lock()
+		defer ji.Unlock()
+		switch e.eventType {
+		case indexEventCollectionAddOrUpdate:
+			ji.indexers[e.collectionName] = e.collectionValue.index(extract)
+		case indexEventCollectionDelete:
+			delete(ji.indexers, e.collectionName)
+		}
+	})
+	return ji
+}
+
+func (j *dynamicjoin[T]) Synced() Syncer {
+	// We use a poll syncer here because the member collections of dynamicjoin can
+	// change; it's not a one-and-done sync check, so channels are inappropriate.
+	return pollSyncer{
+		name: j.collectionName,
+		f:    j.synced.Load,
+	}
+}
+
+func (j *dynamicjoin[T]) HasSynced() bool {
+	return j.syncer.HasSynced()
+}
+
+func (j *dynamicjoin[T]) WaitUntilSynced(stop <-chan struct{}) bool {
+	return j.syncer.WaitUntilSynced(stop)
+}
+
+// The caller MUST have the lock when executing this function
+func (j *dynamicjoin[T]) registerCollectionChangeHandlerLocked(h func(e collectionChangeEvent[T])) {
+	if j.collectionChangeHandlers == nil {
+		j.collectionChangeHandlers = make([]func(collectionChangeEvent[T]), 0)
+	}
+	j.collectionChangeHandlers = append(j.collectionChangeHandlers, h)
+}
+
+// AddOrUpdateCollection adds a collection to the join. When this function returns
+// the passed collection will start influencing event handling
+func (j *dynamicjoin[T]) AddOrUpdateCollection(c Collection[T]) {
+	j.Lock()
+	ic := c.(internalCollection[T])
+	j.collectionsByKey[ic.name()] = ic
+
+	// Add the new collection to the other handler logic
+	for _, h := range j.collectionChangeHandlers {
+		h(collectionChangeEvent[T]{
+			eventType:       indexEventCollectionAddOrUpdate,
+			collectionName:  ic.name(),
+			collectionValue: ic,
+		})
+	}
+	j.Unlock()
+
+	// Change the sync status of our join collection if the new
```

Comment:
This doesn't seem correct. Once a collection is "Synced" its synced permanently.

Synced means "All **initial state** has been processed". So that could be "the initial list passed in, ignoring AddOrUpdateCollection" if we want those to be the semantics. This is how `StaticCollection` _used_ to behave, until I found it helpful to be able to explicitly mark when that is ready too

Reply:
Ack, misremembered the semantics. Easy fix